llms/mlx_lm/utils.py

```diff
@@ -150,20 +200,37 @@ def generate(
            (default ``False``).
        formatter (Optional[Callable]): A function which takes a token and a
            probability and displays it.
+       repetition_penalty (float, optional): The penalty factor for repeating tokens.
+       repetition_context_size (int, optional): The number of tokens to consider for repetition penalty.
     """
 
     if verbose:
         print("=" * 10)
         print("Prompt:", prompt)
 
-    prompt = mx.array(tokenizer.encode(prompt))
+    prompt_tokens = mx.array(tokenizer.encode(prompt))
```

Comment:
Nit: you don't need to change the name of the prompt here do you? Could just keep it as `prompt` as before unless I missed something.

Reply:
it kind of makes sense to change it since the prompt type changed be an array of token ids instead of a string. I'm not sure if I suggested these changes, but I don't have a strong opinion on this..

Reply:
Fair point, let's keep it then.

Reply:
Yes, the repetition penalty needs to be applied on the prompt as well. In my last comment what I was trying to say is that, in the latest approach we are using `repetition_context_size` on the prompt tokens.  

```python
repetition_context = []
     if repetition_penalty and repetition_penalty > 1.0:
         repetition_context = prompt_tokens.tolist()
         repetition_context = repetition_context[-repetition_context_size:] # this part 
```

I am suggesting that we use repetition penalty on the prompt tokens, but not the `repetition_context_size` on the prompt tokens, rather we apply repetition penalty on all the prompt tokens irrespective of the `repetition_context_size`  which will create a window on the prompt. 

Maybe we can use `repetition_context_size` only on the generated tokens. 

Suggested change:

```python
repetition_context = []
     if repetition_penalty and repetition_penalty > 1.0:
         repetition_context = prompt_tokens.tolist()
```

I will setup a PR, it will be easier for review.